tests/epyccel/test_epyccel_variable_annotations.py

```diff
@@ -324,9 +324,41 @@ def dict_empty_init():
 def test_dict_complex_float(language):
     def dict_int_float():
         # Not valid in Python 3.8
-        a : dict[complex, float] #pylint: disable=unsubscriptable-object,unused-variable
+        a : dict[complex, float] #pylint: disable=unsubscriptable-object
         a = {1j:1.0, -1j:2.0}
         return len(a)
 
     epyc_dict_int_float = epyccel(dict_int_float, language = language)
     assert epyc_dict_int_float() == dict_int_float()
+
+def test_inhomogeneous_tuple_annotation_1(language):
+    def inhomogeneous_tuple_annotation():
+        a : tuple[int, bool] = (1, True) #pylint: disable=unsubscriptable-object
+        return a[0], a[1]
+
+    epyc_inhomogeneous_tuple_annotation = epyccel(inhomogeneous_tuple_annotation, language = language)
+    assert epyc_inhomogeneous_tuple_annotation() == inhomogeneous_tuple_annotation()
+
+def test_inhomogeneous_tuple_annotation_2(language):
+    def inhomogeneous_tuple_annotation():
+        a : tuple[int] = (1,) #pylint: disable=unsubscriptable-object
+        return a[0]
+
+    epyc_inhomogeneous_tuple_annotation = epyccel(inhomogeneous_tuple_annotation, language = language)
+    assert epyc_inhomogeneous_tuple_annotation() == inhomogeneous_tuple_annotation()
+
+def test_inhomogeneous_tuple_annotation_3(language):
+    def inhomogeneous_tuple_annotation():
+        a : tuple[int,int,int] = (1,2,3) #pylint: disable=unsubscriptable-object
+        return a[0], a[1], a[2]
+
+    epyc_inhomogeneous_tuple_annotation = epyccel(inhomogeneous_tuple_annotation, language = language)
+    assert epyc_inhomogeneous_tuple_annotation() == inhomogeneous_tuple_annotation()
+
+def test_inhomogeneous_tuple_annotation_4(language):
```

Comment:
The following test raises errors in both `C` and `Fortran`:

```Python
if __name__ == '__main__':
    a : tuple[tuple[int, float]] = ((1,0.2),)
    print(a[0][0], a[0][1])
```

```bash

ERROR at annotation (semantic) stage
pyccel:
 |error [semantic]: boo.py [3,4]| Cannot modify 'const' variable (a)
 ```

Reply:
Fixed over [1f4b1b6](https://github.com/pyccel/pyccel/pull/2066/commits/1f4b1b6a3cd039521b554044618593a3e11e519c) - [bcd4a6c](https://github.com/pyccel/pyccel/pull/2066/commits/bcd4a6c66d3eeb84712dd64b40388611debfa755)

Reply:
Test added here : 413b93ec6

Reply:
Hmm, that's not great. I guess the infinite loop is unrelated. I'll look into it

Reply:
Attempt number 2
Fixed?
e108ec2c5
15de72a1f
6276a382a

Reply:
:sob: You are too good at finding these corner cases :sweat_smile:

Reply:
Fixed : 35886366c4e9cd07b87cdcb7ee342f8ebb161084
with test : 0ad38c9ad360436aaec083ccc3c96c59fa0ca1d8

Reply:
Not quite 😅
```python
if __name__ == '__main__':
    a : tuple[tuple[tuple[tuple[int, float]], int]] = ((((1,0.2),),1),)
```
I think this needs changing:
```python
                    if not isinstance(new_element_type, HomogeneousTupleType):
                        raise_error = True
                        break
```

Reply:
Fixed : d7e6bbd4f
with test : a905aa787

Reply:
Great, this fixes the issue! but you don't try to make tuple homogeneous when possible:
```python
if __name__ == '__main__':
    b : tuple[int,int] = (1,2)
    a : tuple[tuple[tuple[tuple[int, float]], int], tuple[int, int], tuple[float,float]] = ((((1,0.2),),1),b,(0.3,0.2))
```
```C
#include "boo.h"
#include <stdlib.h>
#include <stdint.h>
int main()
{
    int64_t b_0;
    int64_t b_1;
    int64_t a_0_0_0_0;
    double a_0_0_0_1;
    int64_t a_0_1;
    int64_t a_1_0;
    int64_t a_1_1;
    double a_2_0;
    double a_2_1;
    b_0 = INT64_C(1);
    b_1 = INT64_C(2);
    a_0_0_0_0 = INT64_C(1);
    a_0_0_0_1 = 0.2;
    a_0_1 = INT64_C(1);
    a_1_0 = b_0;
    a_1_1 = b_1;
    a_2_0 = 0.3;
    a_2_1 = 0.2;
    return 0;
}
```
Did you have this in mind?

Reply:
I have made this more explicit in the docs: f28c641bd

Reply:
Got it! Awesome